### PR TITLE
Restrict bill max value to prevent rounding error

### LINKED ISSUE #163: [PE-D][Tester E] Bill amount increases by 1 if given a bill above 16 digits long
### Description:

The `bill` command is documented to add to an existing bill a value specified by the user, but a possible floating point error may increase bill amounts incorrectly increasing values by 1 for large inputs. This may potential be dangerous if dealing with a large volume and quantity of transactions as money can appear and disappear erroneously.

### Steps to reproduce:

1. Add a new person and ensure that the total is equal to zero

2. Type `bill 10 b/9999999999999999` or a similar value 16 digits or more (Recommended to reproduce: `9999 9999 9999 9999`)

3. Observe that the bill amount is greater than the value specified. In the image, this value is `1 0000 0000 0000 0000`.

### Suggestion:

1. Check for representational errors in floating point numbers to see if the underlying value is correct

2. Or limit the amount of money that can be added, as this issue doesn't occur for inputs less than 16 digits.

-------------
Labels: `severity.High` `type.FunctionalityBug`
original: rui-han-crh/ped#2

### PR DESCRIPTION
**Pull request by JustinPeng13**
## Added
- Bill test cases for boundary values
- BillCommand check for too large bill + error message

## Modified
- Bill to accept maximum 12 leading digits
- Bill message constraints to include number of digits
- Bill field summary in UG

Resolves #163

### CODE REVIEW

File: docs/UserGuide.md [hunk 1]
```diff
@@ -562,18 +562,17 @@ In the case where GuestBook rejects your field inputs, you can refer to this sec
...
+| **`PHONE`**             | `p/`                 | Should contain only **numbers**, and it should be at least 3 digits and at most 15 digits long.                                                                                                                                                                                                                                                                                                 |
 | **`EMAIL`**             | `e/`                 | Should contain only **alphanumeric characters and 1 `@` character**.<br/>The local segment (before `@`) should be at most 64 characters long, and the domain segment (after `@`) should be at most 255 characters long.<br/>Special characters (`+`, `_`, `.`, `-`) can be included but **not as the first or last characters** of any segment. They also cannot be used adjacent to each other. |
```
Comment (jehousoh) [inline docs/UserGuide.md]: Should we mention the special characters thing at the first sentence? or move the first sentence to the back?

File: docs/UserGuide.md [hunk 2]
```diff
@@ -562,18 +562,17 @@ In the case where GuestBook rejects your field inputs, you can refer to this sec
...
-| **`BILL`**              | `b/`                 | Should contain only **numbers, one optional decimal point (`.`), and one optional leading sign (`+`/`-`)**. <br/>Can have up to 2 decimal places and will be assumed to be positive if no leading sign is used.                                                                                                                                                                                  |
+| **`ROOM`**              | `r/`                 | Should contain only **alphanumeric characters and hyphens (`-`)**.<br/>A hyphen can only be added if it is in between 2 alphanumeric characters.                                                                                                                                                                                                                                                |
```
Comment (jehousoh) [inline docs/UserGuide.md]: Should the prefix for room be rm/?

File: src/test/java/seedu/guest/model/guest/BillTest.java [hunk 3]
```diff
@@ -54,7 +57,8 @@ public void isValidBill() {
         assertTrue(Bill.isValidBill("1.5")); // 1 decimal place
```
Comment (jehousoh) [inline src/test/java/seedu/guest/model/guest/BillTest.java:58]: Should we add decimal point with 1 decimal place? (eg. .9)

**Review by jehousoh — CHANGES_REQUESTED** [review]
Looks good! Left some minor comments.

File: docs/UserGuide.md (hunk 1 above)
Comment (JustinPeng13) [reply docs/UserGuide.md]: Yes, changed

File: docs/UserGuide.md (hunk 2 above)
Comment (JustinPeng13) [reply docs/UserGuide.md]: Yes, changed

File: src/test/java/seedu/guest/model/guest/BillTest.java (hunk 3 above)
Comment (JustinPeng13) [reply src/test/java/seedu/guest/model/guest/BillTest.java:58]: Yes, added

**Review by jehousoh — APPROVED** [review]
lgtm
